silo-to-silo/contracts/src/SiloToSilo.sol

```diff
@@ -0,0 +1,280 @@
+import "@openzeppelin/contracts-upgradeable/security/PausableUpgradeable.sol";
+import "./IEvmErc20.sol";
+
+struct TokenInfo {
+    string nearTokenAccountId;
+    bool isStorageRegistered;
+}
+
+contract SiloToSilo is Initializable, UUPSUpgradeable, AccessControlUpgradeable, PausableUpgradeable {
+    using AuroraSdk for NEAR;
+    using AuroraSdk for PromiseCreateArgs;
+    using AuroraSdk for PromiseWithCallback;
+
+    bytes32 public constant CALLBACK_ROLE = keccak256("CALLBACK_ROLE");
+
+    uint64 constant BASE_NEAR_GAS = 10_000_000_000_000;
+    uint64 constant WITHDRAW_NEAR_GAS = 50_000_000_000_000;
+    uint64 constant FT_TRANSFER_CALL_NEAR_GAS = 150_000_000_000_000;
+    uint128 constant ASCII_0 = 48;
+    uint128 constant ASCII_9 = 57;
+    uint128 constant ONE_YOCTO = 1;
+    uint128 constant NO_DEPOSIT = 0;
+
+    NEAR public near;
+    string public siloAccountId;
+
+    //[auroraErc20Token] => tokenAccountIdOnNear
+    mapping(IEvmErc20 => TokenInfo) public registeredTokens;
+
+    //[auroraErc20Token][userAddressOnAurora] => userBalance
+    mapping(IEvmErc20 => mapping(address => uint128)) public balance;
+
+    event TokenRegistered(IEvmErc20 token, string nearAccountId);
+
+    function initialize(address wnearAddress, string memory _siloAccountId) external initializer {
+        near = AuroraSdk.initNear(IERC20_NEAR(wnearAddress));
+        siloAccountId = _siloAccountId;
+
+        _grantRole(CALLBACK_ROLE, AuroraSdk.nearRepresentitiveImplicitAddress(address(this)));
+        _grantRole(DEFAULT_ADMIN_ROLE, msg.sender);
+    }
+
+    function registerToken(IEvmErc20 token) external {
+        PromiseCreateArgs memory callGetNep141FromErc20 = near.call(
+            siloAccountId,
+            "get_nep141_from_erc20",
+            abi.encodePacked(address(token)),
+            NO_DEPOSIT,
+            BASE_NEAR_GAS
+        );
+
+        PromiseCreateArgs memory callback = near.auroraCall(
+            address(this),
+            abi.encodeWithSelector(this.getNep141FromErc20Callback.selector, token),
+            NO_DEPOSIT,
+            BASE_NEAR_GAS
+        );
+
+        callGetNep141FromErc20.then(callback).transact();
+    }
+
+    function getNep141FromErc20Callback(IEvmErc20 token) external onlyRole(CALLBACK_ROLE) {
+        require(
+            AuroraSdk.promiseResult(0).status == PromiseResultStatus.Successful,
+            "ERROR: The `get_nep141_from_erc20` XCC is fail"
+        );
+
+        string memory nearTokenAccountId = string(AuroraSdk.promiseResult(0).output);
+        registeredTokens[token] = TokenInfo(nearTokenAccountId, false);
+        emit TokenRegistered(token, nearTokenAccountId);
+    }
+
+    function storageDeposit(IEvmErc20 token, uint128 storageDepositAmount) external {
+        string storage tokenAccountId = registeredTokens[token].nearTokenAccountId;
+        require(bytes(tokenAccountId).length > 0, "The token is not registered!");
+
```

Comment:
If `registeredTokens[token].isStorageRegistered == true`, no need call `storage_deposit`. This check can safe some user wNEAR tokens.